### DISCUSSION TITLE
It's possible 'JSON injection' in Prisma data property?

### DISCUSSION THREAD

**Discussion by karimpanacci** [prisma/prisma#21305] [Q&A]
### Question

While developing my code with prisma and express, I noticed a potential security issue that I hadn't noticed.
Let's take an example:

```
export const async (req, res, next) => {
    const id = req.params.id
    const body = req.body

    const accountData = await prisma.userAccount.update({
        where: { id },
        data: {
            active: body.active,
        }
    })
}
```

body.active in theory should be a Boolean value, but it could also be a dictionary... and since prisma sometimes uses options within a dictionary, in your opinion, is it possible for a user, under certain conditions, to write a json that contains particular options that can create unexpected behaviors?

In any case, do you think it is safe to use such code in production? So take the active value directly from the body of the request?

My reasoning is that there is no need to validate the "active" property since prisma will raise an exception in case the data is not of the right type ...

But now I have the doubt that with a particular json and in particular situations, some malicious user can do something with the prisma options within the dictionary

Clearly I know that this "JSON injection" can be done if we pass unsafe data in the property where... this is obvious... but my doubt is specific in the property "data"

**Comment by nurul3101** [comment]
Hello @karim2012 👋 

> In any case, do you think it is safe to use such code in production? So take the active value directly from the body of the request?

I would not suggest you to directly use the `active` parameter from the request body. Even though Prisma has built-in validation for matching the types defined in your schema, it's still a good practice to validate user input at the application layer.

You should explicitly validate that `body.active` is a Boolean before passing it to the query. This can prevent unexpected behaviours and potential security risks. 

You could use `Joi` to validate the request body. It could look something like this:
```ts
import Joi from 'joi';

export const async (req, res, next) => {
    const id = req.params.id
    const body = req.body

    // Define validation schema
    const schema = Joi.object({
        active: Joi.boolean().required(),
    });

    // Validate request body against schema
    const { error } = schema.validate(body);

    if (error) {
        return res.status(400).send(error.details[0].message);
    }

    const accountData = await prisma.userAccount.update({
        where: { id },
        data: {
            active: body.active,
        }
    })
}
```

**Reply by karimpanacci** [reply]
Thank you for the answer, then I will use a validation library... although I think I'll use Typia instead of Joi or Zod... Typia allows me to validate objects directly via the pure typescript types/interface that I will retrieve from the prisma.schema...

So I won't have code duplication...

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by MohammedElMO** [comment]
grap
